x-pack/plugin/esql/arrow/src/main/java/org/elasticsearch/xpack/esql/arrow/BlockConverter.java

```diff
@@ -0,0 +1,477 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.arrow;
+
+import org.apache.arrow.memory.ArrowBuf;
+import org.apache.arrow.vector.types.Types;
+import org.apache.arrow.vector.types.pojo.FieldType;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.breaker.NoopCircuitBreaker;
+import org.elasticsearch.common.io.stream.RecyclerBytesStreamOutput;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.compute.data.Block;
+import org.elasticsearch.compute.data.BlockFactory;
+import org.elasticsearch.compute.data.BooleanBlock;
+import org.elasticsearch.compute.data.BytesRefBlock;
+import org.elasticsearch.compute.data.DoubleBlock;
+import org.elasticsearch.compute.data.IntBlock;
+import org.elasticsearch.compute.data.LongBlock;
+
+import java.io.IOException;
+import java.util.BitSet;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+public abstract class BlockConverter {
+
+    private final FieldType fieldType;
+    private final String esqlType;
+
+    protected BlockConverter(String esqlType, Types.MinorType minorType) {
+        // Add the exact ESQL type as field metadata
+        var meta = Map.of("elastic:type", esqlType);
+        this.fieldType = new FieldType(true, minorType.getType(), null, meta);
+        this.esqlType = esqlType;
+    }
+
+    public final String esqlType() {
+        return this.esqlType;
+    }
+
+    public final FieldType arrowFieldType() {
+        return this.fieldType;
+    }
+
+    protected final <T extends Block> T castBlock(Block block, Class<T> clazz) {
+        if (clazz.isAssignableFrom(block.getClass())) {
```

Comment:
This is pretty much what doing `(DoubleBlock) block` does but it works with the runtime `Class` object. I think it's probably ok to just do the cast without the method.

Reply:
Removed in c77c2459e2b7cec24793b5a664f42615f281d1dd to use a regular cast. This was useful in development, but this should never fail as unit tests test all field types.